onnx/common/path.h

```diff
@@ -8,17 +8,67 @@
 #pragma once
 
 #include <string>
+#ifdef _WIN32
+#include <windows.h>
+#endif
 
 namespace ONNX_NAMESPACE {
 
 #ifdef _WIN32
 const std::string k_preferred_path_separator = "\\";
+const std::wstring w_k_preferred_path_separator = L"\\";
 #else // POSIX
 const std::string k_preferred_path_separator = "/";
 #endif
 
-std::string path_join(const std::string& origin, const std::string& append);
-void normalize_separator(std::string& path);
-std::string clean_relative_path(const std::string& path);
+template <typename STRING>
+STRING path_join(const STRING& origin, const STRING& append);
+
+template <typename CHAR>
+bool is_path_separator(CHAR c) {
```

Comment:
You should not need to have this function for Windows. Instead, you should iterate `path` as a container. 

Like this:
```C++
// filesystem_path_example.cpp
// compile by using: /EHsc /W4 /permissive- /std:c++17 (or later)
#include <string>
#include <iostream>
#include <sstream>
#include <filesystem>

using namespace std;
using namespace std::filesystem;

wstring DisplayPathInfo()
{
    // This path may or may not refer to an existing file. We are
    // examining this path string, not file system objects.
    path pathToDisplay(L"C:/FileSystemTest/SubDir3/SubDirLevel2/File2.txt ");

    wostringstream wos;
    int i = 0;
    wos << L"Displaying path info for: " << pathToDisplay << endl;
    for (path::iterator itr = pathToDisplay.begin(); itr != pathToDisplay.end(); ++itr)
    {
        wos << L"path part: " << i++ << L" = " << *itr << endl;
    }

    wos << L"root_name() = " << pathToDisplay.root_name() << endl
        << L"root_path() = " << pathToDisplay.root_path() << endl
        << L"relative_path() = " << pathToDisplay.relative_path() << endl
        << L"parent_path() = " << pathToDisplay.parent_path() << endl
        << L"filename() = " << pathToDisplay.filename() << endl
        << L"stem() = " << pathToDisplay.stem() << endl
        << L"extension() = " << pathToDisplay.extension() << endl;

    return wos.str();
}

int main()
{
    wcout << DisplayPathInfo() << endl;
    // wcout << ComparePaths() << endl; // see following example
    wcout << endl << L"Press Enter to exit" << endl;
    wstring input;
    getline(wcin, input);
}
```